src/Renci.SshNet/Security/Cryptography/Ciphers/AesGcmCipher.cs

```diff
@@ -0,0 +1,101 @@
+﻿#if NETSTANDARD2_1_OR_GREATER || NETCOREAPP3_0_OR_GREATER
+using System;
+using System.Buffers.Binary;
+using System.Security.Cryptography;
+
+using Renci.SshNet.Common;
+
+namespace Renci.SshNet.Security.Cryptography.Ciphers
+{
+    /// <summary>
+    /// AES GCM cipher implementation.
+    /// <see href="https://datatracker.ietf.org/doc/html/rfc5647"/>.
+    /// </summary>
+    public sealed class AesGcmCipher : AeadCipher, IDisposable
+    {
+        private readonly byte[] _nonce;
+        private readonly AesGcm _aesGcm;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AesGcmCipher"/> class.
+        /// </summary>
+        /// <param name="key">The key.</param>
+        /// <param name="iv">The IV.</param>
+        public AesGcmCipher(byte[] key, byte[] iv)
+            : base(key, tagSize: 16)
+        {
+            _nonce = iv.Take(12);
+#if NET8_0_OR_GREATER
+            _aesGcm = new AesGcm(key, TagSize);
+#else
+            _aesGcm = new AesGcm(key);
+#endif
+        }
+
+        /// <inheritdoc/>
+        public override byte[] Encrypt(byte[] input, int offset, int length)
+        {
+            // [outbound sequence][packet length field][padding length field sz][payload][random paddings]
+            // [--4 bytes(offset)][------4 bytes------][-------------------Plain Text--------------------]
+            var packetLengthField = new ReadOnlySpan<byte>(input, offset, 4);
+            var plainText = new ReadOnlySpan<byte>(input, offset + 4, length - 4);
+
+            var result = new byte[length + TagSize];
+            packetLengthField.CopyTo(result);
+            var cipherText = new Span<byte>(result, 4, length - 4);
+            var tag = new Span<byte>(result, length, TagSize);
+
+            _aesGcm.Encrypt(_nonce, plainText, cipherText, tag, packetLengthField);
+
+            IncrementCounter();
+
+            return result;
+        }
+
+        /// <inheritdoc/>
+        public override byte[] Decrypt(byte[] input, int offset, int length)
+        {
+            // [inbound sequence][packet length field][padding length field sz][payload][random paddings][Authenticated TAG]
+            // [-----4 bytes----][----4 bytes(offset)][------------------Cipher Text--------------------][-------TAG-------]
+            var packetLengthField = new ReadOnlySpan<byte>(input, offset - 4, 4);
```

Comment:
I understand your concern. Typically, the cipher only decrypts "cipher text" which starts from offset and ends before MAC (tag). However, `AesGcm` requires the packet_length (aad) and MAC (tag) to do the decryption which is **before** and **after** the cipher text.
The `Decrypt(...)` method of `AesGcm` is called (and only called) from https://github.com/sshnet/SSH.NET/blob/develop/src/Renci.SshNet/Session.cs#L1302 which guarantees the data layout.

Reply:
> The `Decrypt(...)` method of `AesGcm` is called (and only called) from https://github.com/sshnet/SSH.NET/blob/develop/src/Renci.SshNet/Session.cs#L1302 which guarantees the data layout.

Still, adding a bounds check could be a good thing for future proofing.

Reply:
There are some options here:
1. keep it as is but add bounds check
2. in `Session` class, treat aead cipher as special and set offset before packet_length and include MAC(tag) in length
3. in `Session` class, treat aead cipher as special and set offset to 0 and include all in length (consider that ChaCha20Poly1305 requires sequence block if I read the [spec](https://datatracker.ietf.org/doc/html/draft-josefsson-ssh-chacha20-poly1305-openssh-00) correctly)

Reply:
Just checked the source code of the constructor about `ReadOnlySpan<T>`. It does check the bounds:

- It casts the `start` to ulong/uint to make sure it is positive.
- It checks the `length` to no larger than array length - offset.

```
        /// <summary>
        /// Creates a new read-only span over the portion of the target array beginning
        /// at 'start' index and ending at 'end' index (exclusive).
        /// </summary>
        /// <param name="array">The target array.</param>
        /// <param name="start">The index at which to begin the read-only span.</param>
        /// <param name="length">The number of items in the read-only span.</param>
        /// <remarks>Returns default when <paramref name="array"/> is null.</remarks>
        /// <exception cref="System.ArgumentOutOfRangeException">
        /// Thrown when the specified <paramref name="start"/> or end index is not in the range (&lt;0 or &gt;Length).
        /// </exception>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public ReadOnlySpan(T[]? array, int start, int length)
        {
            if (array == null)
            {
                if (start != 0 || length != 0)
                    ThrowHelper.ThrowArgumentOutOfRangeException();
                this = default;
                return; // returns default
            }
#if TARGET_64BIT
            // See comment in Span<T>.Slice for how this works.
            if ((ulong)(uint)start + (ulong)(uint)length > (ulong)(uint)array.Length)
                ThrowHelper.ThrowArgumentOutOfRangeException();
#else
            if ((uint)start > (uint)array.Length || (uint)length > (uint)(array.Length - start))
                ThrowHelper.ThrowArgumentOutOfRangeException();
#endif

            _pointer = new ByReference<T>(ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(array), (nint)(uint)start /* force zero-extension */));
            _length = length;
        }
```

Reply:
> There are some options here:
> 
> 1. keep it as is but add bounds check
> 2. in `Session` class, treat aead cipher as special and set offset before packet_length and include MAC(tag) in length
> 3. in `Session` class, treat aead cipher as special and set offset to 0 and include all in length (consider that ChaCha20Poly1305 requires sequence block if I read the [spec](https://datatracker.ietf.org/doc/html/draft-josefsson-ssh-chacha20-poly1305-openssh-00) correctly)

Personally I prefer option1, because it is consistent for all kinds of ciphers. No special handing in `Session` class. The current implementation in https://github.com/sshnet/SSH.NET/pull/1369 is option2. It defers the issue when add `ChaCha20Poly1305` in the future.

Life will be much easier if there's no such many options :upside_down_face:

Reply:
Changed back to option1 with some tiny enhancements in the latest commit.